frame/dvm/ethereum/src/lib.rs

```diff
@@ -285,21 +286,21 @@ pub mod pallet {
 
 			let extracted_transaction = match transaction {
 				Transaction::Legacy(t) => Ok(Transaction::Legacy(ethereum::LegacyTransaction {
-					nonce: darwinia_evm::Pallet::<T>::account_basic(&source).nonce, // auto set
-					gas_price: T::FeeCalculator::min_gas_price(),                   // auto set
+					nonce,                                        // auto set
+					gas_price: T::FeeCalculator::min_gas_price(), // auto set
 					gas_limit: t.gas_limit,
 					action: t.action,
 					value: t.value,
 					input: t.input,
 					signature: t.signature, // not used.
 				})),
-				_ => Err(Error::<T>::MessageTransactionError),
+				_ => Err(Error::<T>::InvalidTransactionType),
 			}?;
 
-			ensure!(
-				Self::validate_transaction_in_block(source, &extracted_transaction).is_ok(),
-				Error::<T>::MessageValidateError
-			);
+			if let Err(err) = Self::validate_transaction_in_block(source, &extracted_transaction) {
+				Self::deposit_event(Event::MessageTransactValidateRejected { source, nonce, err });
```

Comment:
Can't we tell The reason using the Error since Error is also enum?

Could we add struct to MessageValidateRejected.

This way of adding event looks weird.

Reply:
> Can't we tell The reason using the Error since Error is also enum?
> Could we add struct to MessageValidateRejected.

As for as I know, the `Pallet Error` does not support struct type now, so the `MessageValidateRejected` can not contain the more detailed message. Another way is to create many runtime error types, then needs to convert the validated error type to the corresponding runtime error type.